File: dsaWithJava/tree/binarysearchtree/BSTRecursive.java
package tree.binarysearchtree;

public class BSTRecursive {
    Node root;

    public void insert(int data) {
        root = insert(root, data);
    }

    public void delete(int data) {
        root = delete(root, data);
    }

    public boolean search(int data) {
        return search(root, data);
    }

    public void display() {
        display(root);
        System.out.println();
    }

    // Time Complexity -> O(h)
    // Auxiliary Space -> O(h)
    private Node insert(Node root, int data) {
        if (root == null)
            return new Node(data);
        if (root.data > data)
            root.left = insert(root.left, data);
        else if (root.data < data)
            root.right = insert(root.right, data);
        return root;
    }

    // Time Complexity -> O(h)
    // Auxiliary Space -> O(h)
    private Node delete(Node root, int data) {
        if (root == null)
            return null;
        if (root.data > data)
            root.left = delete(root.left, data);
        else if (root.data < data)
            root.right = delete(root.right, data);
        else {
            if (root.left == null)
                return root.right;
            else if (root.right == null)
                return root.left;
            Node successor = getSuccessor(root);
            root.data = successor.data;
            root.right = delete(root.right, root.data);
        }
        return root;
    }

    private Node getSuccessor(Node root) {
        Node currentNode = root.right;
        while (currentNode != null && currentNode.left != null)
            currentNode = currentNode.left;
        return currentNode;
    }

    // Time Complexity -> O(h)
    // Auxiliary Space -> O(h)
    private boolean search(Node root, int data) {
        if (root == null)
            return false;
        if (root.data == data)
            return true;
        if (root.data > data)
            return search(root.left, data);
        else
            return search(root.right, data);
    }

    // Time Complexity -> O(n)
    // Auxiliary Space -> O(h)
    private void display(Node root) {
        if (root == null)
            return;
        display(root.left);
        System.out.print(root.data + " ");
        display(root.right);
    }
}

File: Python/tree/binary_search_tree/bst_recursive.py
from node import Node

root = None


def insert(data):
    global root

    def __insert(root: Node, data) -> Node:
        if root is None:
            return Node(data)
        if root.data > data:
            root.left = __insert(root.left, data)
        elif root.data < data:
            root.right = __insert(root.right, data)
        return root

    root = __insert(root, data)


def delete(data):
    global root

    def __delete(root: Node, data) -> Node:
        if root is None:
            return None
        if root.data > data:
            root.left = __delete(root.left, data)
        elif root.data < data:
            root.right = __delete(root.right, data)
        else:
            if root.left is None:
                return root.right
            elif root.right is None:
                return root.left
            root.data = __get_successor(root)
            root.right = __delete(root.right, root.data)
        return root

    def __get_successor(root: Node) -> int:
        current_node = root.right
        while current_node is not None and current_node.left is not None:
            current_node = current_node.left
        return current_node.data

    root = __delete(root, data)


def search(data) -> bool:
    global root

    def __search(root: Node, data) -> bool:
        if root is None:
            return False
        if root.data == data:
            return True
        if root.data > data:
            return __search(root.left, data)
        else:
            return __search(root.right, data)

    return __search(root, data)


def display():
    global root

    def __display(root: Node):
        if root is None:
            return
        __display(root.left)
        print(root.data, end=" ")
        __display(root.right)

    __display(root)
    print()
